Add log for security-related events

Mats-SX opened the pull request:
Replaces #7817 

changelog: Log all security related events to the new security.log file

Commit:
Added security log and logged login attempts

Mats-SX committed:
Add logging for creation and deletion of users

Mats-SX committed:
Minor refactorings

Mats-SX committed:
Add logging for adding roles to users

Mats-SX committed:
Add acceptance scenario for security logging

Mats-SX committed:
Log removal of roles from users

Mats-SX committed:
Log password changes

Mats-SX committed:
Move security log setting to `SecuritySettings`

Mats-SX committed:
Log unauthorized password change attempts

- Expose `SecurityLog` through procedure context in enterprise deployments

fickludd committed:
Move logging from user manager to procedures

Mats-SX committed:
Add tests for logging unauthorized operations

Mats-SX committed:
Add tests for logging suspension of users

Mats-SX committed:
Add tests for logging activation of users

Mats-SX committed:
Refactor user and role name validation

Clean up login logging testing to support new log format

Mats-SX committed:
Add test for failed login attempts due to bad auth token

Mats-SX committed:
Add tests for logging creation of roles

Mats-SX committed:
Add tests for logging deletion of roles

Mats-SX committed:
Add tests for failed attempts to list users and roles

fickludd committed:
Added logging of changePassword calls

- also included password_change_required state where appropriate

fickludd committed:
Redirected LdapRealm logs to security log

fickludd committed:
Exposed setting to control whether successful authentications are logged or not

fickludd committed:
Introduced rotation of security logs and made logging asynchronous

henriknyman commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/LdapRealm.java:
```diff
@@ -417,14 +425,22 @@ private String parseLdapServerUrl( String rawLdapServer )
-                        ((SearchResult) result.next()).toString() ) );
+                if ( securityLog.isDebugEnabled() )
+                {
+                    securityLog.warn( withRealm(
```
This is warning is pretty serious. I wonder if it will easily be too easily missed if we require isDebugEnabled() since that is not the default. I think it is better to leave it as a warning even though it is a configuration problem.

Mats-SX replied:
Are you suggesting that we simply remove the if-case and always log this as a warning?

henriknyman commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/LdapRealm.java:
```diff
@@ -448,28 +463,31 @@ private String parseLdapServerUrl( String rawLdapServer )
-        return proceedWithSearch;
+        if ( !proceedWithSearch )
+        {
+            throw new IllegalArgumentException( "Illegal LDAP user search settings, see security log for details." );
```
If we now throw exception then maybe we should change warnings to errors for those cases

fickludd committed:
Added security log to Ha, Core and Edge modules

henriknyman commented on enterprise/core-edge/src/main/java/org/neo4j/coreedge/core/EnterpriseCoreEditionModule.java:
```diff
@@ -121,6 +126,18 @@ public void registerProcedures( Procedures procedures )
+        {
+            securityLog = new SecurityLog( config, fileSystem,
+                    jobScheduler.executor( JobScheduler.Groups.internalLogRotation ) );
+        }
```
This little piece of lazy creation seem to be duplicated across all enterprise editions. Maybe we could extract a shared helper.

Mats-SX replied:
I tried this, but I don't see how it's possible, since there isn't a common superclass with access to the enterprise `SecurityLog` that can access the field.

henriknyman commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/EditionModule.java:
```diff
@@ -117,19 +130,39 @@ public static AuthManager createAuthManager( Config config, LogService logging )
+            logging.getInternalLog( GraphDatabaseFacadeFactory.class ).warn( "Unable to create log for auth-manager. " +
+                    "Auth logging turned off." );
+            authManagerLog = NullLog.getInstance();
+        }
```
This could also be a `createSecurityLog` (`createAuthManagerLog`) method, and then we can just give the log it returns as a parameter to `createAuthManager`.
(Although I suspect we may soon need both dependencies (file system, job scheduler) in auth manager anyways, comment is about as this PR stands on its own)

Mats-SX replied:
I agree that it would simplify the signature of this method, but it would make calling this method more cumbersome, as you'd need to pass two parameters twice (`config` and `logging`) to the both respective create methods. The caller would then also get access to the log, which is not intended.

Mats-SX committed:
Code cleanups

Mats-SX committed:
Extract log creation to private helper

fickludd committed:
Refactored auth log creation

fickludd committed:
FIxed broken tests

fickludd committed:
Reverted some refactoring in security procedures tests

fickludd committed:
Minor fixes after rebase, and clean-up of imports

fickludd committed:
Refactored procedure service finding

- ensure that enterprise-auth-procedures are not loaded when testing community databases

fickludd committed:
Disabled slf4j redirect to neo4j debug log

fickludd committed:
Ignore failures due to missing procedure providers

fickludd committed:
Made EnterpriseAuthManagerFactory more robust toward missing SecurityLog

fickludd committed:
Refactor of createAuthManager and clarification in Neo4jWithSocket
